circleguard/visualizer.py

```diff
@@ -143,59 +144,60 @@ def next_frame(self):
-
-            # remove old hitobjects
-            running_hitobjs = []
-            for hitobj in self.hitobjs:
-                if hitobj.type == "slider":
-                    if hitobj.time+(hitobj.slider_length*hitobj.slider_repeats) > current_time:
-                        running_hitobjs.append(hitobj)
-                if hitobj.type == "circle":
-                    if hitobj.time > current_time:
-                        running_hitobjs.append(hitobj)
-                if hitobj.type == "spinner":
-                    if hitobj.spinner_length > current_time:
-                        running_hitobjs.append(hitobj)
-            self.hitobjs = running_hitobjs
-
-            if self.beatmap.last_time < current_time:
+            self.get_hitobjects()
+            if self.beatmap.hitobjects[-1].time+3000 < current_time:
                 self.reset()
         self.update_signal.emit(current_time)
         self.update()
 
+    def get_hitobjects(self):
+        # calc preempt, fade_in, hitwindow
+        if self.beatmap.difficulty["ApproachRate"] == 5:
+            preempt = 1200
+            fade_in = 800
+        elif self.beatmap.difficulty["ApproachRate"] < 5:
+            preempt = 1200 + 600 * (5 - self.beatmap.difficulty["ApproachRate"]) / 5
+            fade_in = 800 + 400 * (5 - self.beatmap.difficulty["ApproachRate"]) / 5
+        else:
+            preempt = 1200 - 750 * (self.beatmap.difficulty["ApproachRate"] - 5) / 5
+            fade_in = 800 - 500 * (self.beatmap.difficulty["ApproachRate"] - 5) / 5
+        hitwindow = 150 + 50 * (5 - self.beatmap.difficulty["OverallDifficulty"]) / 5
+
+        # get current hitobjects
+        time = self.clock.get_time()
+        found_all = False
+        index = 0
+        self.hitobjs = []
+        while not found_all:
+            current_hitobj = self.beatmap.hitobjects[index]
+            if current_hitobj.time-preempt < time < current_hitobj.time or ((current_hitobj.time-preempt < time < current_hitobj.time + current_hitobj.duration) if 2 & current_hitobj.type else False):  # 
+                current_hitobj.preempt = preempt
+                current_hitobj.fade_in = fade_in
+                current_hitobj.hitwindow = hitwindow
+                self.hitobjs.append(current_hitobj)
+            elif current_hitobj.time-preempt > time and not 2 & current_hitobj.type:
+                found_all = True
+            index += 1
+            if index == len(self.beatmap.hitobjects)-1:
+                found_all = True
+
+
     def paintEvent(self, event):
         """
         Called whenever self.update() is called. Draws all cursors and Hitobjects
         """
         painter = QPainter(self)
         painter.setRenderHint(QPainter.Antialiasing)
-        if self.beatmap_path != "":
-            self.paint_beatmap(painter)
+        painter.setPen(PEN_WHITE if (get_setting("dark_theme") or get_setting("visualizer_bg")) else PEN_BLACK)
```

Comment:
call get_setting("visualizer_bg") once and save result

Reply:
for the record I wouldn't be worried about this micro optimization if it wasn't occuring in such a tight loop

Reply:
Are you talking about ln 192 and how it should only trigger if "visualizer_bg" changed? I need to call get get_setting("visualizer_bg") every loop or you won't be able to change the bg on the fly.

Reply:
calling it every paintEvent is fine, but it should only be called once per paint event, currently it's called twice